File: java/src/main/java/ai/synapsenetwork/sdk/SynapseProvider.java
package ai.synapsenetwork.sdk;

import java.util.List;
import java.util.Map;

public final class SynapseProvider {
  private final SynapseAuth auth;

  SynapseProvider(SynapseAuth auth) {
    this.auth = auth;
  }

  public SynapseAuth.IssueProviderSecretResult issueSecret(SynapseAuth.CredentialOptions options) {
    return auth.issueProviderSecret(options);
  }

  public List<SynapseAuth.ProviderSecret> listSecrets() {
    return auth.listProviderSecrets();
  }

  public SynapseAuth.ProviderSecretDeleteResult deleteSecret(String secretId) {
    return auth.deleteProviderSecret(secretId);
  }

  public SynapseAuth.ProviderRegistrationGuide getRegistrationGuide() {
    return auth.getRegistrationGuide();
  }

  public SynapseAuth.ServiceManifestDraft parseCurlToServiceManifest(String curlCommand) {
    return auth.parseCurlToServiceManifest(curlCommand);
  }

  public SynapseAuth.RegisterProviderServiceResult registerService(SynapseAuth.RegisterProviderServiceOptions options) {
    return auth.registerProviderService(options);
  }

  public SynapseAuth.RegisterProviderServiceResult registerLlmService(SynapseAuth.RegisterProviderServiceOptions options) {
    options.serviceKind = "llm";
    options.priceModel = "token_metered";
    return auth.registerProviderService(options);
  }

  public List<SynapseAuth.ProviderServiceRecord> listServices() {
    return auth.listProviderServices();
  }

  public SynapseAuth.ProviderServiceRecord getService(String serviceId) {
    return auth.getProviderService(serviceId);
  }

  public SynapseAuth.ProviderServiceStatus getServiceStatus(String serviceId) {
    return auth.getProviderServiceStatus(serviceId);
  }

  public SynapseAuth.ProviderServiceUpdateResult updateService(String serviceRecordId, Map<String, Object> patch) {
    return auth.updateProviderService(serviceRecordId, patch);
  }

  public SynapseAuth.ProviderServiceDeleteResult deleteService(String serviceRecordId) {
    return auth.deleteProviderService(serviceRecordId);
  }

  public SynapseAuth.ProviderServicePingResult pingService(String serviceRecordId) {
    return auth.pingProviderService(serviceRecordId);
  }

  public SynapseAuth.ProviderServiceHealthHistory getServiceHealthHistory(String serviceRecordId, int limit) {
    return auth.getProviderServiceHealthHistory(serviceRecordId, limit);
  }

  public SynapseAuth.ProviderEarningsSummary getEarningsSummary() {
    return auth.getProviderEarningsSummary();
  }

  public SynapseAuth.ProviderWithdrawalCapability getWithdrawalCapability() {
    return auth.getProviderWithdrawalCapability();
  }

  public SynapseAuth.ProviderWithdrawalIntentResult createWithdrawalIntent(
      String amountUsdc, String idempotencyKey, String destinationAddress) {
    return auth.createProviderWithdrawalIntent(amountUsdc, idempotencyKey, destinationAddress);
  }

  public SynapseAuth.ProviderWithdrawalList listWithdrawals(int limit) {
    return auth.listProviderWithdrawals(limit);
  }
}

File: dotnet/src/SynapseNetwork.Sdk/SynapseProvider.cs
namespace SynapseNetwork.Sdk;

public sealed class SynapseProvider(SynapseAuth auth)
{
    public Task<IssueProviderSecretResult> IssueSecretAsync(CredentialOptions? options = null, CancellationToken cancellationToken = default)
        => auth.IssueProviderSecretAsync(options, cancellationToken);

    public Task<IReadOnlyList<ProviderSecret>> ListSecretsAsync(CancellationToken cancellationToken = default)
        => auth.ListProviderSecretsAsync(cancellationToken);

    public Task<ProviderSecretDeleteResult> DeleteSecretAsync(string secretId, CancellationToken cancellationToken = default)
        => auth.DeleteProviderSecretAsync(secretId, cancellationToken);

    public Task<ProviderRegistrationGuide> GetRegistrationGuideAsync(CancellationToken cancellationToken = default)
        => auth.GetRegistrationGuideAsync(cancellationToken);

    public Task<ServiceManifestDraft> ParseCurlToServiceManifestAsync(string curlCommand, CancellationToken cancellationToken = default)
        => auth.ParseCurlToServiceManifestAsync(curlCommand, cancellationToken);

    public Task<RegisterProviderServiceResult> RegisterServiceAsync(RegisterProviderServiceOptions options, CancellationToken cancellationToken = default)
        => auth.RegisterProviderServiceAsync(options, cancellationToken);

    public Task<RegisterProviderServiceResult> RegisterLlmServiceAsync(RegisterProviderServiceOptions options, CancellationToken cancellationToken = default)
    {
        options.ServiceKind = "llm";
        options.PriceModel = "token_metered";
        return auth.RegisterProviderServiceAsync(options, cancellationToken);
    }

    public Task<IReadOnlyList<ProviderServiceRecord>> ListServicesAsync(CancellationToken cancellationToken = default)
        => auth.ListProviderServicesAsync(cancellationToken);

    public Task<ProviderServiceRecord> GetServiceAsync(string serviceId, CancellationToken cancellationToken = default)
        => auth.GetProviderServiceAsync(serviceId, cancellationToken);

    public Task<ProviderServiceStatus> GetServiceStatusAsync(string serviceId, CancellationToken cancellationToken = default)
        => auth.GetProviderServiceStatusAsync(serviceId, cancellationToken);

    public Task<ProviderServiceUpdateResult> UpdateServiceAsync(string serviceRecordId, IReadOnlyDictionary<string, object?> patch, CancellationToken cancellationToken = default)
        => auth.UpdateProviderServiceAsync(serviceRecordId, patch, cancellationToken);

    public Task<ProviderServiceDeleteResult> DeleteServiceAsync(string serviceRecordId, CancellationToken cancellationToken = default)
        => auth.DeleteProviderServiceAsync(serviceRecordId, cancellationToken);

    public Task<ProviderServicePingResult> PingServiceAsync(string serviceRecordId, CancellationToken cancellationToken = default)
        => auth.PingProviderServiceAsync(serviceRecordId, cancellationToken);

    public Task<ProviderServiceHealthHistory> GetServiceHealthHistoryAsync(string serviceRecordId, int limit = 100, CancellationToken cancellationToken = default)
        => auth.GetProviderServiceHealthHistoryAsync(serviceRecordId, limit, cancellationToken);

    public Task<ProviderEarningsSummary> GetEarningsSummaryAsync(CancellationToken cancellationToken = default)
        => auth.GetProviderEarningsSummaryAsync(cancellationToken);

    public Task<ProviderWithdrawalCapability> GetWithdrawalCapabilityAsync(CancellationToken cancellationToken = default)
        => auth.GetProviderWithdrawalCapabilityAsync(cancellationToken);

    public Task<ProviderWithdrawalIntentResult> CreateWithdrawalIntentAsync(string amountUsdc, string? idempotencyKey = null, string? destinationAddress = null, CancellationToken cancellationToken = default)
        => auth.CreateProviderWithdrawalIntentAsync(amountUsdc, idempotencyKey, destinationAddress, cancellationToken);

    public Task<ProviderWithdrawalList> ListWithdrawalsAsync(int limit = 100, CancellationToken cancellationToken = default)
        => auth.ListProviderWithdrawalsAsync(limit, cancellationToken);
}
